config/remote/service: clarify locking behaviors

ajwerner opened the pull request:
### What does this PR do?

This PR is a bundle of cleanups around the code in `pkg/config/remote/service`. It fixes some race conditions, and more generally attempts to tame and document the use of mutexes around this service.

### Motivation

I'm making changes to this package as part of https://datadoghq.atlassian.net/browse/DEBUG-4590 and was having a hard time getting a handle of the synchronization.

### Describe how you validated your changes

All the tests pass, and no behavior should be changed.

### Additional Notes

This pattern of using an anonymous struct with an embedded mutex to document what is protected by that mutex is a pattern I've seen widely used in the Go ecosystem and one I've found broadly beneficial.

*PLEASE REVIEW COMMIT BY COMMIT*

ajwerner committed:
config/remote/service: remove unused abstraction

This `Service` structure is exported but with no users and now no methods.
Furthermore, it embeds a mutex and two fields, but the two fields that it
embeds don't actually use that mutex -- they are never modified.

Instead we hoist the fields up to the `CoreAgentService`, the actual user
of these fields.

ajwerner committed:
config/remote/service: remove unused receiver

ajwerner committed:
config/remote/service: unexport Mutex

The embedding before made `Lock()` and `Unlock()` part of the public
API of the `CoreAgentService`. This was not intentional.

ajwerner committed:
config/remote/service: clarify Mutex usage

ajwerner committed:
config/remote/service: make CoreAgentService.Stop idempotent

The comment said it should be, but nobody ever set this channel to nil,
so subsequent calls would panic.

ajwerner committed:
config/remote/service: sprinkle some mutex TODOs

ajwerner committed:
config/remote/service: drop mutex while sleeping

mellon85 commented on pkg/config/remote/service/clients.go:
```diff
@@ -23,29 +23,57 @@ type cacheBypassClients struct {
-	windowDuration time.Duration
-	capacity       int
-	allowance      int
+	mu struct {
```
this struct is never called by multiple goroutines, while this adds close to no overhead it's not necessary

ajwerner committed:
config/remote/service: refactor refresh bypass rate limiting

Rename `cacheBypassClients` to `rateLimiter` and extract the bypass request
channel from the rate limiter to CoreAgentService, clarifying the threading
model and separation of concerns.

ajwerner reviewed (commented):
@mellon85 thanks for the review!

ajwerner committed:
config/remote/service: improve lock management in refresh

ajwerner committed:
config/remote/service: no need for mutex around api.API

This dependency is never changed.

ajwerner reviewed (commented):
@mellon85 thank you for the review! I added 3 commits to the end to address the feedback. Let me know if I missed anything

ajwerner commented:
/gitlab trigger-ci

ajwerner committed:
config/remote/service: restructured OrgPoller

Isolate the state of the orgpoller to clarify synchronization.

ajwerner commented on pkg/config/remote/service/service.go:
```diff
@@ -126,70 +114,81 @@ func (s *Service) getTargetFiles(uptaneClient uptaneClient, targetFilePaths []st
+		sync.Mutex
 
-	agentVersion string
+		db     *bbolt.DB
```
Fine by me, I'm happy to rebase your change.

mellon85 commented on pkg/config/remote/service/service.go:
```diff
@@ -652,120 +748,86 @@ func logRefreshError(s *CoreAgentService, err error) {
 		log.Warnf("[%s] could not get previous TUF version state: %v", s.rcType, err)
 	}
-	if s.forceRefresh() || err != nil {
+	if s.forceRefreshLocked() || err != nil {
```
as it's used only once and accesses a field covered by mutex I'd remove it
